### PR TITLE
machine.mem: Add mem functions into machine module

### PR DESCRIPTION
**Pull request by dhylands**
This factors out the mem8/16/32 functions from extmod/modmachine.c
and stmhal/modstm.c and creates extmod/machine_mem.[ch]

modmachine.c was added to the unix port
stmhal/modmachine.c and stmhal/modstm.c both use extmod/machine_mem
When desired, modstm can be removed from the stmhal port.

make-stmconst.py was renamed to make-memconst.py. No changes
were made to the script itself.

I tested on stmhal using the following to turn on and off the
GREEN LED.

> > > machine.mem32[machine.GPIOA + machine.GPIO_ODR] = 0x4000
> > > machine.mem32[machine.GPIOA + machine.GPIO_ODR] = 0
> > > machine.mem32[0x40020014] = 0x4000
> > > machine.mem32[0x40020014] = 0

This patch (with modmachine and modstm both having mem) adds 2030 bytes.
This patch and removing modstm reduces the size by 60 bytes.

### CODE REVIEW

File: extmod/machine_mem.c [hunk 1]
```diff
@@ -3,7 +3,7 @@
  *
  * The MIT License (MIT)
  *
- * Copyright (c) 2013, 2014 Damien P. George
+ * Copyright (c) 2013, 2014, 2015 Damien P. George
```
Comment (stinos) [inline extmod/machine_mem.c]: I've been wondering about this for a while: is the date really necessary (because copyright isn't timeless or so?), and is the plan to continue to just add years?

Comment (dhylands) [inline extmod/machine_mem.c]: Perhaps not. I did a little digging and came up with this:
http://www.contentious.com/2007/01/07/copyright-notice-is-the-year-really-necessary/

I think its really a call for @dpgeorge to make.

**Comment by dhylands** [comment]
I realized that by making stm module be an alias for machine module, then we get our 2000 bytes back and still maintain backwards compatability.

**Comment by pfalcon** [comment]
This patch shows typical problem we have with contributions - instead of submitting one clear patch which does single unquestionable thing, which the can be merged immediately, people put many things in one patch, some of them being questionable, thus blocking progress of the entire patch. I'm really interested to know why it's easier for people to spend half an hour today and then a month in wait, instead of 5-10 today, and then next 5-10 minutes tomorrow (or the other day), etc.

Back to specifics, there was no requirement for machine.mem* to replace stm.mem*. Actually, with explicit remark of @dpgeorge that became anti-requirement - machine.mem* should not try to replace stm.mem*. It's also easy to see why that's the case - because they have conflicting requirements: machine.mem wants to provide ability to access entire physical address space without imposing and even assuming any restrictions, while stm.mem* wants to provide optimized access to subset (not even full set) of STM32 registers, while imposing various adhoc access restrictions.

Then, I don't think that stm should be ever gone completely - it's nice place to host STM32-specific constants.

**Comment by dhylands** [comment]
> This patch shows typical problem we have with contributions - instead of submitting one clear patch which does single unquestionable thing, which the can be merged immediately, people put many things in one patch, some of them being questionable, thus blocking progress of the entire patch. I'm really interested to know why it's easier for people to spend half an hour today and then a month in wait, instead of 5-10 today, and then next 5-10 minutes tomorrow (or the other day), etc.

I don't read minds. So instead, I put together my idea what should be done, and propose it. Then we have a discussion about it. I make changes based on the feedback and submit it again.

> Back to specifics, there was no requirement for machine.mem to replace stm.mem. Actually, with explicit remark of @dpgeorge that became anti-requirement - machine.mem* should not try to replace stm.mem. 

I disagree. I think that having machine.mem and stm.mem would be totally confusing.

> It's also easy to see why that's the case - because they have conflicting requirements: machine.mem wants to provide ability to access entire physical address space without imposing and even assuming any restrictions, while stm.mem wants to provide optimized access to subset (not even full set) of STM32 registers, while imposing various adhoc access restrictions.

Actually, I think you got the requirement wrong. The current stm.mem does provide access to the entire physical memory space. The only optimization that stm.mem makes is that it allows a certain subset of that space to be accessed from interrupt handlers.

This patch adds a configurable access restriction - if you don't want it - don't configure it. I'm not quite sure what you mean by "optimized" access. The only thing that modstm does that the unix machine.mem doesn't was to allow some negative addresses to be mapped to some positive addresses, which allows interrupt handlers to access memory using small ints. The way that this patch is coded, that functionality is retained.

Maybe what you're really trying to say is that you don't like the fact that I masked the address with 0x7fffffff because it doesn't suit your accesses. Which is fine. I didn't realize that you used that codepath, so I can make it configurable. But it would have been alot more constructive to make a line comment where I put that and tell me, rather than this big comment that you made.

> Then, I don't think that stm should be ever gone completely - it's nice place to host STM32-specific constants.

So, to summarize your comment, I think what you're saying is that the code to access the memory should be in machine.mem and the constants should stay in stm.

I'm definitely not going to write a patch which provides machine.mem and stm.mem because I think that's plain confusing. I'd rather remove machine.mem and say that memory access goes into a port specific module.

And rather than me guess about what you're objecting to, it would be much more constructive to make a comment on some lines of code and tell me what you don't like about instead of going these grand tirades which I don't find particularly useful.

**Comment by dhylands** [comment]
> This is not valid for most of the CPUs out there. It's nothing but an STM32-specific hack. It can be put into generic modmachine source code if you want, but should be clearly #ifdef'ed as STM32-specific hack (which has limits of its usage even there, like not being able to access standard Cortex-M control block at 0xexxxxxxx).

You can access the standard Corete-M control block at 0xexxxxxxx by providing a integer address of 0xexxxxxxx. The mask with 0x7fffffff only changes how negative addresses are interpreted. And it has a side effect on 64-bit small int address which I didn't intend (my goof).

I agree with you, I goofed here. I think that this masking should be a port specific thing. and I'll change it to be that way.

**Comment by pfalcon** [comment]
> I don't read minds.

And find more and more need to read contributors' minds. Because we have growing set of "unmergeable" pull requests. It's not too grave problem, consider open vs merged requests ration of 37/867, but nonetheless, number of unmerged PRs growing, and as a maintainer (i.e. someone who can merge PRs, and thus having a responsibility to do that), I'm concerned with how to make it easier for everyone. And I really wonder why people err on the side of making complex, rather than simple, changes.

> I disagree. I think that having machine.mem and stm.mem would be totally confusing.

That fact that you disagree is a good criteria for separating patches related to machine.mem from patches related to stm.mem (first can be merged right away, second need more discussion).

> I agree with you, I goofed here.

What I'm trying to hint is that I see direct correlation between complexity of a patch and few unrelated things being packed into it and goofs like that. Again, I wonder why contributors don't consider this side of things.

> it doesn't suit your accesses. Which is fine. I didn't realize that you used that codepath,

It's not about me. It's just I'm taking a role of caring about generic solutions to problems (physical memory access in this case), which would work for any platform (or as many as possible platforms), whereas most other people care about single piece of hardware.

> so I can make it configurable. 

You can and we can. As a separate patch, please. First patch would rather be simple and clean. Configurability, optimizations, hacks can be added as separate patches (after separate discussion).

> But it would have been alot more constructive to make a line comment where I put that and tell me, rather than this big comment that you made.

That's actually what I did, because it was very easy to see that "& 0x7fffffff" can't be correct in the general case. But again, I'm not less interested to know how it happens that people submit patches with such issues, and how it make easier for everyone to avoid that (because a maintainer can miss some problematic change and ship broken thing to many people). As you see, I consider simplicity of each separate commit, then each pull request to be a key. I wonder if other people disagree with me, or just don't give it a thought/user other criteria when making patches (I wonder which then?)

**Comment by dpgeorge** [comment]
A large open source project is difficult to run and maintain, and difficult to coordinate, from both the contributors and maintainers side.  As such we should go a little easier on ourselves and try to have a more friendly conversation.  It's not the end of the world if something is wrong or breaks, that's what CI and tagged releases are for.

**Comment by pfalcon** [comment]
Yeah, and then before "taking it easy" (== starting to ignore) I'm really trying to understand how other people make patches. If we can discuss a particular patch, why can't we discuss more higher-level process regarding them? It's not completely random whim after processing close to 1K of them. That said, it of course makes sense to do if there're interest, if people think it's waste of time or worse, then well...

**Comment by dhylands** [comment]
I'm going to close this and redo it as a bunch of separate PRs as suggested by @pfalcon.
The PRs I'll do are:
1 - Factor extmod/modmachine.c into extmod/machine_mem.c and unix/modmachine.c
2 - Add machine_mem functions into stmhal/modmachine.c
If we want to move stm register constants into machine (not sure if this is desirable or not)
   3 - Move modstm constants to modmachine 
   4 - Alias stm to machine (to provide backwards compatability and save flash)
else
   3 - Point stm.mem functions to machine_mem (and remove stm.mem functions)
